[ftxui] Add new port (fix #16327)

myd7349 opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #16327

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  <all / linux, windows, ...>, <Yes/No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Your answer`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  <Yes / I am still working on this PR>

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

myd7349 committed:
[ftxui] Add new port (fix #16327)

myd7349 committed:
[ftxui] x-add-version ftxui

ArthurSonzogni commented on ports/ftxui/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+  "description": "C++ Functional Terminal User Interface",
+  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
+  "license": "MIT",
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
+  ]
```
Out of curiosity, what are those about?

JonLiu1993 replied:
When using vcpkg_cmake_config_fixup(), you need to add vcpkg-cmake-config, and use other new functions only need to add vcpkg-cmake

ArthurSonzogni reviewed (commented):
Thanks for working on this!
I wanted to start doing the same, but since the work has already been done here, better doing the same.

I see this uses version-date.
At some point, I would like to be able to adhere to semantic versioning, when I will have reached 1.0.
Do you think, this will still be possible later to switch to it?

myd7349 commented:
Hi! @ArthurSonzogni Thanks for your awesome work.
Yeah. I'd like to switch to v1.0 if it was released in the near future.

MichaelGoulding commented on ports/ftxui/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
+  "license": "MIT",
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
```
```suggestion
```

MichaelGoulding replied:
I had to remove this to get it to work for me.

JackBoosY replied:
It always be the host dependency.

JonLiu1993 commented:
@myd7349 ,Thanks for your pr.
Could you take a look?
[install-x64-uwp-dbg-out.log](https://github.com/microsoft/vcpkg/files/6659703/install-x64-uwp-dbg-out.log)

myd7349 committed:
[ftxui] Drop UWP support

According to the build log:
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(56,37): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(60,32): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
ftxui can not be built successfully on UWP, since `getenv` is unavailable on UWP.

https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160

myd7349 committed:
[ftxui] Update to 2021-06-16

myd7349 committed:
[ftxui] Overwrite version

myd7349 committed:
Merge branch 'master' into ftxui-init

myd7349 commented:
> 
> 
> @myd7349 ,Thanks for your pr.
> Could you take a look?
> [install-x64-uwp-dbg-out.log](https://github.com/microsoft/vcpkg/files/6659703/install-x64-uwp-dbg-out.log)

Hi! @JonLiu1993 Thanks for your review.

According to the build log:
> D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(56,37): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(60,32): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]

ftxui can not be built successfully on UWP, since `getenv` is unavailable on UWP:
https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160

myd7349 commented:
Hi! @MichaelGoulding That was strange. Could you verify your `vcpkg-tool` version?
> ./vcpkg version

JonLiu1993 commented on ports/ftxui/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
```
```suggestion
vcpkg_fail_port_install(ON_TARGET "uwp") 
```

JonLiu1993 commented on versions/f-/ftxui.json:
```diff
@@ -0,0 +1,14 @@
+      "version-date": "2021-06-16",
+      "port-version": 0
+    },
+    {
+      "git-tree": "4f720eb108815befe381654803dbdae18ee0615d",
+      "version-date": "2021-06-09",
+      "port-version": 0
+    }
```
Please remove this

myd7349 committed:
[ftxui] Try to fix linux and macOS build

myd7349 committed:
[ftxui] Overwrite version

myd7349 committed:
[ftxui] Fail fast on UWP

myd7349 committed:
[ftxui] Clean version file

myd7349 committed:
[ftxui] Overwrite version

myd7349 commented:
Hi! @JonLiu1993 Thanks for your review.

myd7349 commented:
/ci hurry up!

myd7349 committed:
[ftxui] Update to 2021-06-17

myd7349 committed:
[ftxui] Overwrite version

myd7349 committed:
[ftxui] Remove debug/share

myd7349 committed:
[ftxui] Overwrite version

myd7349 commented:
One more note:
https://github.com/microsoft/vcpkg/issues/16327#issuecomment-863945954
https://github.com/microsoft/vcpkg/issues/16327#issuecomment-865455700

JackBoosY commented on ports/ftxui/vcpkg.json:
```diff
@@ -0,0 +1,18 @@
+  "description": "C++ Functional Terminal User Interface",
+  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
+  "license": "MIT",
+  "supports": "!uwp",
```
Is the upstream doesn't support uwp or it doesn't support uwp in vcpkg currently?

JackBoosY commented on ports/ftxui/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" "${CURRENT_PACKAGES_DIR}/debug/share")
+
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)

```

myd7349 commented:
Hi! @*JackBoosY*

According to the build log:
>
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(56,37):
error C2039: 'getenv': is not a member of 'std'
[D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(60,32):
error C2039: 'getenv': is not a member of 'std'
[D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]

ftxui can not be built successfully on UWP, since getenv is unavailable on
UWP:
https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160

Jack·Boos·Yu ***@***.***> 于2021年6月22日周二 下午3:26写道：

> *@JackBoosY* commented on this pull request.
> ------------------------------
>
> In ports/ftxui/vcpkg.json
> <https://github.com/microsoft/vcpkg/pull/18360#discussion_r655947800>:
>
> > @@ -0,0 +1,18 @@
> +{
> +  "name": "ftxui",
> +  "version-date": "2021-06-17",
> +  "description": "C++ Functional Terminal User Interface",
> +  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
> +  "license": "MIT",
> +  "supports": "!uwp",
>
> Is the upstream doesn't support uwp or it doesn't support uwp in vcpkg
> currently?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/vcpkg/pull/18360#pullrequestreview-689149712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJPCAI4UKHQWM62VOGDRJTTUA3LFANCNFSM46NIP5TQ>
> .
>

myd7349 committed:
Merge branch 'master' into ftxui-init

myd7349 committed:
Merge branch 'ftxui-init' of https://github.com/myd7349/vcpkg into ftxui-init

myd7349 committed:
[ftxui] Quote paths

myd7349 committed:
[ftxui] Overwrite version

myd7349 commented:
Hi! @JackBoosY Thanks. Updated.
